rts/motoko-rts/src/stabilization/graph_copy/limit.rs

```diff
@@ -0,0 +1,32 @@
+use crate::stabilization::ic0_performance_counter;
+
+/// Instruction limit for the graph copy increment.
+/// Monitoring the message instruction counter.
+pub struct InstructionLimit {
+    start: u64,
+    limit: u64,
+}
+
+impl InstructionLimit {
+    pub fn new(limit: u64) -> InstructionLimit {
+        InstructionLimit {
+            start: Self::instruction_counter(),
+            limit,
+        }
+    }
+
+    pub fn is_exceeded(&self) -> bool {
+        let current = Self::instruction_counter();
+        debug_assert!(self.start <= current);
+        let elapsed = current - self.start;
+        elapsed > self.limit
+    }
+
+    pub fn reset(&mut self) {
+        self.start = Self::instruction_counter();
+    }
+
+    fn instruction_counter() -> u64 {
+        unsafe { ic0_performance_counter(0) }
```

Comment:
This might actually be quite expensive to call frequently (IIRC like 200 instructions per call?)

Reply:
Right, I will measure the overhead incurred. Maybe we need a more coarse-grained counter.

Reply:
Indeed, this can be quite expensive, like 2 billion instructions for an array of length of 10 million elements. I will try to implement an improvement based on synthetic estimated counting with occasional calibration to the real costs.

Reply:
I optimzed it, by calling `ic0_performance_counter` less frequently, only when a certain amount of objects have been copied **or** a certain amount of memory has been processed. 

Together with some more tuning (chunk size of memory clearing), I obtain the following improvements:

### Upgrade Instructions with Incremental Graph Copy

Benchmark      | Before Optimization        | After Optimization         | Difference
---------------|----------------------------|----------------------------|-----------------
Double Sharing | 690_000                    | 260_000                    | -62%
Linked List    | 86_000_000                 | 71_000_000                 | -17%     
Array List     | 5_800_000_000              | 3_400_000_000              | -41%
Small Objects  | 52_000_000_000             | 36_000_000_000             | -31%
Small Texts    | 20_000_000_000             | 15_000_000_000             | -25%
Big Numbers    | 4_100_000_000              | 3_100_000_000              | -24%
Blobs          | 940_000_000                | 950_000_000                | +1%

(The "Blobs" are little bit slower due to another recent change, where I made the stable memory clearing incremental, see below.)